### DISCUSSION TITLE
Model nomic-embed-text:v1.5 does not exist and could not be pulled

### DISCUSSION THREAD

**Discussion by Verfinix** [nilsherzig/LLocalSearch#71] [Q&A]
I am able to connect ollama using
http://0.0.0.0:11434/
and return
Ollama is running

but having error below, anyone know how to resolve this ?

Model nomic-embed-text:v1.5 does not exist and could not be pulled: Post "http://0.0.0.0:11434/api/pull": dial tcp 0.0.0.0:11434: connect: connection refused

**Comment by nilsherzig** [comment]
Localhost inside the backend container isn't the localhost of your host PC. You have to find the network interface the backend is using, which IP your host device has on this interface and input this IP into the ollama URL environment variable. 

Let me know how it goes and if you need more help.

**Reply by Verfinix** [reply]
Can I know where should I configure this

Model nomic-embed-text:v1.5 does not exist and could not be pulled: Post "http://0.0.0.0:11434/api/pull": dial tcp 0.0.0.0:11434: connect: connection refused

in .env I have configured


OLLAMA_HOST=http://172.16.0.201:11434
MAX_ITERATIONS=30
CHROMA_DB_URL=http://chromadb:8000
SEARXNG_DOMAIN=http://searxng:8080
SEARXNG_HOSTNAME=localhost

Where else required to configure the ollama host ?

**Reply by veeragoni** [reply]
i get the same error. interestingly. it says about minlm even though I have latest code that supposed to pull nomic embedding model.

Edit:
I pulled the containers and now i get this error

`Model nomic-embed-text:v1.5 does not exist and could not be pulled: Post "http://localhost:11434/api/pull": dial tcp [::1]:11434: connect: connection refused`

`Model all-minilm:v2 does not exist and could not be pulled: Post "http://localhost:11434/api/pull": dial tcp [::1]:11434: connect: connection refused`

my ollama is running at http://localhost:11434 fine (its actually working with [open-webui](https://github.com/open-webui/open-webui) on 3000. so, only change I did is to map 3001 instead of 3000 in Dockerfile.

But i think thats not the issue here. its somehow not able to communicate with the ollama even though other projects like above are working fine.

**Reply by Verfinix** [reply]
Just to share, I too run open-webui on port 3000, and I map 3003 to 4173 in docker-compose.yaml file

Like below, does this cause a problem ?

  frontend:
    depends_on:
      - backend
    image: nilsherzig/llocalsearch-frontend:latest
    ports:
      - '3003:4173'
    networks:
      - llm_network

**Comment by Verfinix** [comment]
I have also changed the IP to the IP address of the Ollama server in docker file, but it still trying to access from http://0.0.0.0:11434

Model nomic-embed-text:v1.5 does not exist and could not be pulled: Post "http://0.0.0.0:11434/api/pull": dial tcp 0.0.0.0:11434: connect: connection refused 

Below is from the docker-compose.yaml file

services:
  backend:
    image: nilsherzig/llocalsearch-backend:latest
    environment:
      - OLLAMA_HOST=${OLLAMA_HOST:-http://172.16.66.201:11434}
      - CHROMA_DB_URL=${CHROMA_DB_URL:-http://chromadb:8000}
      - SEARXNG_DOMAIN=${SEARXNG_DOMAIN:-http://searxng:8080}
      - MAX_ITERATIONS=${MAX_ITERATIONS:-30}

Where else needed to be configure ?

**Reply by veeragoni** [reply]
let me know if you were able to fix the issue.

**Reply by TheNinthWave** [reply]
This fixed the issue for me on linux: https://dev.to/mjnaderi/accessing-host-services-from-docker-containers-1a97#:~:text=For%20instance%2C%20you%20might%20have,network_mode%3A%20host%20in%20Docker%20Compose.

Add this to your YAML at the top:

database-relay:
    image: alpine/socat:latest
    network_mode: host
    command: TCP-LISTEN:11434,fork,bind=host.docker.internal TCP-CONNECT:127.0.0.1:11434
    extra_hosts:
      - host.docker.internal:host-gateway

**Reply by TheNinthWave** [reply]
Ironically, I asked Perplexity this question and it gave me the same answer as option 3 of three provided options  (the other options being the ones that worked for others). Here is Perplexity's response:

3. Use socat to forward the connection:

If the host service is listening on 127.0.0.1 (localhost) inside the host machine, you can use the socat utility to forward the connection.
In your Docker Compose file, add a service that uses socat to forward the connection:

yaml

services:
  database-relay:
    image: alpine/socat:latest
    network_mode: host
    command: TCP-LISTEN:5432,fork,bind=host.docker.internal TCP-CONNECT:127.0.0.1:5432
    extra_hosts:
      - "host.docker.internal:host-gateway"
  my-service:
    image: some-image
    command: my-command --connect-to host.docker.internal:5432
    extra_hosts:
      - "host.docker.internal:host-gateway"

This will set up a database-relay service that listens on host.docker.internal:5432 and forwards the connection to 127.0.0.1:5432 on the host machine.

Your my-service container can then connect to host.docker.internal:5432 to access the host's database service.

**Reply by Verfinix** [reply]
Have added above in docker YAML file
but still not working.
Notice the log
Attaching to backend-1, chromadb-1, database-relay-1, frontend-1, redis-1, searxng-1
database-relay-1  | 2024/04/10 13:13:17 socat[1] E bind(5, {AF=2 192.168.65.2:11434}, 16): Address not available
database-relay-1 exited with code 1

and after check in local host

netstat -nl |grep 11434

tcp6       0      0 :::11434                :::*                    LISTEN 

notice that the ollama only listen in IP6 and not IP4, look like is a ollama issue instead.

Waiting for ollama feedback on IP4 issue

**Reply by TheNinthWave** [reply]
Can you post the YAML file you are using? Are you using this IP in this line?

command: TCP-LISTEN:11434,fork,bind=host.docker.internal TCP-CONNECT:127.0.0.1:11434

**Comment by nilsherzig** [comment]
> I am able to connect ollama using http://0.0.0.0:11434/ and return Ollama is running
> 
> but having error below, anyone know how to resolve this ?
> 
> Model nomic-embed-text:v1.5 does not exist and could not be pulled: Post "http://0.0.0.0:11434/api/pull": dial tcp 0.0.0.0:11434: connect: connection refused

Hi, sorry for the late reply, but localhost inside the backend docker isn't your hosts localhost. The container has its own loopback device. Check the new compose files to see a working example with a new ollama instance inside the right docker network or connect to your host using the right network interface (not loopback).

**Reply by Verfinix** [reply]
I managed to resolved it by setting the IP to a static IP, and export Ollama to static IP, and is ok now.
